ambari-server/src/main/java/org/apache/ambari/server/controller/internal/BlueprintConfigurationProcessor.java

```diff
@@ -1472,31 +1568,35 @@ String updateForClusterCreate(String propertyName,
                                              ClusterTopology topology);
   }
 
-  private static class HostGroupUpdater implements PropertyUpdater {
+  static class HostGroupUpdater implements PropertyUpdater {
 
-    public static final PropertyUpdater INSTANCE = new HostGroupUpdater();
+    static final HostGroupUpdater INSTANCE = new HostGroupUpdater();
 
     @Override
     public String updateForClusterCreate(String propertyName,
       String origValue,
       Map<String, Map<String, String>> properties,
       ClusterTopology topology) {
 
-      //todo: getHostStrings
-      Matcher m = HostGroup.HOSTGROUP_REGEX.matcher(origValue);
-      if (m.find()) {
-        String hostGroupName = m.group(1);
+      HostGroups hostGroups = new HostGroups(topology);
 
-        HostGroupInfo groupInfo = topology.getHostGroupInfo().get(hostGroupName);
-        if (groupInfo == null) {
-          //todo: this should be validated in configuration validation
-          throw new RuntimeException(
-            "Encountered a host group token in configuration which couldn't be matched to a host group: "
-              + hostGroupName);
-        }
+      //todo: getHostStrings (?)
 
-        //todo: warn if > hosts
-        return origValue.replace(m.group(0), groupInfo.getHostNames().iterator().next());
+      // replaces all %HOSTGROUP::name% references to host names in the value string one by one. The value string can contain
+      // only one (typical) or multiple %HOSTGROUP references. If the same host group is referenced multiple times,
+      // another host will be picked each time
+      // Assuming you have the following hostgroups and hosts:
+      // - hostgroup1: [group1_host]
+      // - hostgroup2: [group2_host]
+      // - hostgroup3: [group3_host1, group3_host2, group3_host3]
+      // the following replacements will be made:
+      // - %HOSTGROUP::group1%:8080 --> grop1_host:8080
+      // - %HOSTGROUP::group1%:8080,%HOSTGROUP::group2%:8080 --> group1_host:8080,group2_host:8080
+      // - %HOSTGROUP::group3%:8080,%HOSTGROUP::group3%:8080,%HOSTGROUP::group3%:8080 -->
+      //      group3_host1:8080,group3_host2:8080,group3_host3:8080 (maybe in different order)
+      for (Matcher m = HostGroup.HOSTGROUP_REGEX.matcher(origValue); m.find(); ) {
+        origValue = m.replaceFirst(hostGroups.nextHost(m.group(1)));
+        m = HostGroup.HOSTGROUP_REGEX.matcher(origValue);
```

Comment:
Is there a possibility that this could result in an infinite loop?  

For example: If a HOSTGROUP token existed in a property that did not match to one of the host group names defined in the Blueprint, then "origValue" might never get updated, causing the regex matcher to keep detecting the HOSTGROUP token that was never replaced.  

Can you please investigate this, to make sure this isn't a possibility?  If it is, then maybe change the loop conditions to guarantee that an infinite loop won't occur in the even of a mis-configured Blueprint.

Reply:
Good catch. It actually could result in an infinite loop in the unlikely case when a host group is named "%HOSTGROUP::something%. I rewrote the replacement logic to take care of that.